File: tests/examples/pointer.py
from examples.utils import *

def case_pointer():
    data = read_json('pointer.cpp.json')
    expect_eq(data['Array']['fields'][0]['types'][0]['type'] , 'int')
    expect_true(data['Array']['fields'][0]['isReduced'])
    expect_true(data['Array']['fields'][0]['types'][0]['isReduced'])
    expect_eq(data['ClzArray']['fields'][0]['types'][0]['type'] , 'Foo')
    expect_eq(data['BasicPointer']['fields'][0]['types'][0]['type'] , 'int')
    expect_true(data['BasicPointer']['fields'][0]['isReduced'])
    expect_true(data['BasicPointer']['fields'][0]['types'][0]['isReduced'])
    expect_eq(data['ClzPointer']['fields'][0]['types'][0]['type'] , 'Foo')
    expect_false(data['ClzPointer']['fields'][0]['isReduced'])
    expect_false(data['ClzPointer']['fields'][0]['types'][0]['isReduced'])
    expect_eq(data['PointerArray']['fields'][0]['types'][0]['type'] , 'Foo')
    expect_eq(data['Reference']['fields'][0]['types'][0]['type'] , 'Foo')

File: tests/examples/pointer.cpp
class Foo {
public:
    Foo() {}
    int a_int;
};

class Array {
    int a[10];
};

class ClzArray {
    Foo foo[10];
};

class BasicPointer {
    int *foo;
};

class ClzPointer {
    Foo *foo;
};

class PointerArray {
    Foo *foo[10];
};

class Reference {
    Foo &foo;

public:
    explicit Reference(Foo &foo) : foo(foo) {}
};
